### DISCUSSION TITLE
2.0 Migration of joinedload from strings: how to handle deeper relationships

### DISCUSSION THREAD

**Discussion by jvanasco** [sqlalchemy/sqlalchemy#10073] [Usage Questions] [answered]
Regarding this 2.0 migration element: https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#orm-query-joining-loading-on-relationships-uses-attributes-not-strings

I have an existing query that is roughly in the form of:

```!python
     .options(
        subqueryload("foo"),
        joinedload("foo.bar"),
        joinedload("foo.bar.biz"),
```

What is the correct way to migrate this to 2.0?  I assumed it might be:

```!python
     .options(
        subqueryload(My.foo),
        joinedload(My.foo.bar),
        joinedload(My.foo.bar.biz),
```
but that brings up comparator and attribute errors.  I've been playing with the inspector, but not much luck.

**Comment by zzzeek** [comment]
each parent class, options joined in a chain on the dot:  **edit: fixed**

```
  .options(
        subqueryload(My.foo).
        joinedload(Foo.bar).
        joinedload(Bar.biz)
   )

```

**Reply by jvanasco** [reply]
Thanks Mike. 

I had tried that too, but got the following error.  Does it look like I missed an obvious migration step based on the partial model below?

I'm clearly not migrating correctly.

```
sqlalchemy.exc.ArgumentError: Mapped class Mapper[CertificateSigned(certificate_signed)] does not apply to any of the root entities in this query, e.g. Mapper[Domain(domain)]. Please specify the full path from one of the root entities to the target attribute. 
```

The model is roughly:

```!Python
    class Domain(Base):
        __tablename__ = "domain"
        id = Column(Integer, primary_key=True)
        certificate_signed_id__latest_single = Column(
            Integer, ForeignKey("certificate_signed.id"), nullable=True
        )
        certificate_signed__latest_single = relationship(
            "CertificateSigned",
            primaryjoin="Domain.certificate_signed_id__latest_single==CertificateSigned.id",
            uselist=False,
        )
    class CertificateSigned(Base):
        __tablename__ = "certificate_signed"
        id = Column(Integer, primary_key=True)
        private_key_id = Column(
            Integer, ForeignKey("private_key.id"), nullable=False
        )
        private_key = relationship(
            "PrivateKey",
            primaryjoin="CertificateSigned.private_key_id==PrivateKey.id",
            uselist=False,
            back_populates="certificate_signeds",
        )
    class PrivateKey(Base):
        __tablename__ = "private_key"
        id = sa.Column(sa.Integer, primary_key=True)
```

and the select is:

```
    q = dbSession.query(Domain).filter(Domain.id == domain_id)\
        .options(
            subqueryload(Domain.certificate_signed__latest_single),
            joinedload(CertificateSigned.private_key),
        )
    item = q.first()
```

**Reply by zzzeek** [reply]
does it get any different if you use selectinload instead of subqueryload ?   I dont see what would cause that problem from what is shown (**edit: I am not reading carefully enough**)

**Reply by jvanasco** [reply]
Sadly, no.

I'll make a reproducible test case, and hopefully that will surface a mistake in my code somewhere.  If not, I'll share it here.

**Reply by zzzeek** [reply, marked as answer]
OH nevremind, you are using a comma, not a period, and so am I, fix:

```py
  .options(
        subqueryload(My.foo).
        joinedload(Foo.bar).
        joinedload(Bar.biz)
  )
```

**Reply by jvanasco** [reply]
Bingo! Thanks, Mike!

**Reply by jvanasco** [reply]
@zzzeek if you have time for one more question...

In 1.4 string syntax I can generate a fairly wide join that maps multiple relationship paths.

```
  .options(
        subqueryload("foo").
        joinedload("foo.a").
        joinedload("foo.b").
        joinedload("foo.c").
        joinedload("foo.c.d")
  )
```

I can't seem to get this working in the new syntax, as the joinedloads try to join against the previous clause - so selecting a or b will keep me from doing c and c.d.

I could do this with additional subqueryloads, but I have a lot of code to port and would like to minimize changes.

**Reply by zzzeek** [reply]
there is a [sub-option syntax](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#specifying-sub-options-with-load-options) and this time I think you mean commas for those first two:

```py
.options(
        subqueryload(A.foo).options(
            joinedload(Foo.a),
           joinedload(Foo.b),
          joinedload(Foo.c).joinedload(C.d)
        )
  )
```

**Reply by jvanasco** [reply]
I was actually looking at the suboptions syntax on the same file , and it never occurred to me to use it for this migration. 

thanks again, mike!
